FR #1646: Standardize Normative Language: Convert All RECOMMENDED Statements to SHOULD

jpradocueva opened the pull request:
As suggested in FR #1646, replace normative statements RECOMMENDED with SHOULD

jpradocueva committed:
replace RECOMMENDED with SHOULD

shawnalpay commented on specification/attributes/data_generator_calculated_split_cost_allocation_handling.md:
```diff
@@ -31,7 +31,7 @@ An attribute that allows data generators to offer more detailed cost and usage i
 * A FOCUS dataset MAY contain records for the unused or unallocated usage from the origin charge as separate allocated charges, if it aligns to the data generator's documented allocation method.
 * Allocated charge records MAY contain apportioned costs for the unused or unallocated usage from the origin charge, if it aligns to the data generator's documented allocation method.
-* Split cost allocation is RECOMMENDED to be applied to charges on an opt-in basis.
+* Split cost allocation is SHOULD to be applied to charges on an opt-in basis.
```
```suggestion
* Split cost allocation SHOULD be applied to charges on an opt-in basis.
```

kk09v commented on specification/datasets/cost_and_usage/columns/availabilityzone.md:
```diff
@@ -6,7 +6,7 @@ An [*availability zone*](#glossary:availability-zone) is a host-provider-assigne
 AvailabilityZone adheres to the following requirements:
 
-* AvailabilityZone is RECOMMENDED to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) when the host provider supports deploying resources or services within an *availability zone*.
+* AvailabilityZone is SHOULD to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) when the host provider supports deploying resources or services within an *availability zone*.
```
```suggestion
* AvailabilityZone SHOULD be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) when the host provider supports deploying resources or services within an *availability zone*.
```

kk09v commented on specification/datasets/cost_and_usage/columns/chargefrequency.md:
```diff
@@ -6,7 +6,7 @@ Charge Frequency indicates how often a [*charge*](#glossary:charge) will occur.
 ChargeFrequency adheres to the following requirements:
 
-* ChargeFrequency is RECOMMENDED to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
+* ChargeFrequency is SHOULD to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
```
```suggestion
* ChargeFrequency SHOULD be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
```

kk09v commented on specification/datasets/cost_and_usage/columns/invoiceid.md:
```diff
@@ -6,7 +6,7 @@ An Invoice ID is a invoice-issuer-assigned identifier for an invoice encapsulati
 InvoiceId adheres to the following requirements:
 
-* InvoiceId is RECOMMENDED to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
+* InvoiceId is SHOULD to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
```
```suggestion
* InvoiceId SHOULD be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
```

kk09v commented on specification/datasets/cost_and_usage/columns/pricingcurrencycontractedunitprice.md:
```diff
@@ -8,7 +8,7 @@ PricingCurrencyContractedUnitPrice adheres to the following requirements:
 * PricingCurrencyContractedUnitPrice presence in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
   * PricingCurrencyContractedUnitPrice MUST be present in a Cost and Usage *FOCUS dataset* when the service provider supports prices in virtual currency and publishes unit prices exclusive of discounts.
-  * PricingCurrencyContractedUnitPrice is RECOMMENDED to be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
+  * PricingCurrencyContractedUnitPrice is SHOULD to be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
```
```suggestion
  * PricingCurrencyContractedUnitPrice SHOULD be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
```

kk09v commented on specification/datasets/cost_and_usage/columns/pricingcurrencyeffectivecost.md:
```diff
@@ -8,7 +8,7 @@ PricingCurrencyEffectiveCost adheres to the following requirements:
 * PricingCurrencyEffectiveCost presence in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
   * PricingCurrencyEffectiveCost MUST be present in a Cost and Usage *FOCUS dataset* when the service provider supports prices in virtual currency and publishes unit prices exclusive of discounts.
-  * PricingCurrencyEffectiveCost is RECOMMENDED to be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
+  * PricingCurrencyEffectiveCost is SHOULD to be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
```
```suggestion
  * PricingCurrencyEffectiveCost SHOULD be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
```

kk09v commented on specification/datasets/cost_and_usage/columns/pricingcurrencylistunitprice.md:
```diff
@@ -8,7 +8,7 @@ PricingCurrencyListUnitPrice adheres to the following requirements:
 * PricingCurrencyListUnitPrice presence in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset) is defined as follows:
   * PricingCurrencyListUnitPrice MUST be present in a Cost and Usage *FOCUS dataset* when the service provider supports prices in virtual currency and publishes unit prices exclusive of discounts.
-  * PricingCurrencyListUnitPrice is RECOMMENDED to be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
+  * PricingCurrencyListUnitPrice is SHOULD to be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
```
```suggestion
  * PricingCurrencyListUnitPrice SHOULD be present in a Cost and Usage *FOCUS dataset* when the service provider supports pricing and billing in different currencies and publishes unit prices exclusive of discounts.
```

kk09v commented on specification/datasets/cost_and_usage/columns/servicesubcategory.md:
```diff
@@ -6,7 +6,7 @@ The Service Subcategory is a secondary classification of the [Service Category](
 ServiceSubcategory adheres to the following requirements:
 
-* ServiceSubcategory is RECOMMENDED to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
+* ServiceSubcategory is SHOULD to be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
```
```suggestion
* ServiceSubcategory SHOULD be present in a Cost and Usage [*FOCUS dataset*](#glossary:FOCUS-dataset).
```

shawnalpay commented:
@ijurica would like @jpradocueva to update the guidelines as well.

shawnalpay reviewed (changes requested):
Incorporate @kk09v's suggestions and revise the guidelines per @ijurica.

jpradocueva committed:
Update specification/attributes/data_generator_calculated_split_cost_allocation_handling.md

Co-authored-by: Shawn Alpay <shawn@finops.org>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/availabilityzone.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/chargefrequency.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/invoiceid.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/pricingcurrencycontractedunitprice.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/pricingcurrencyeffectivecost.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/pricingcurrencylistunitprice.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
Update specification/datasets/cost_and_usage/columns/servicesubcategory.md

Co-authored-by: Karl <133434112+kk09v@users.noreply.github.com>

jpradocueva committed:
updates editorial guidelines by removing RECOMMENDED

jpradocueva commented:
@ijurica, I have updated the editorial guidelines by removing the usage of RECOMMENDED. See [commit](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/1647/commits/8d251f467f50e5b552195f257dd8101804ff614f). Could you please check what you think? Thanks

shawnalpay commented:
Discussed in Dec 1 Maintainers.  Need to touch the [normative requirements guidelines](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/blob/working_draft/guidelines/normative-requirements-guidelines.md) to not delineate between SHOULD and RECOMMENDED.  Setting this PR to draft until this is done.

shawnalpay commented on guidelines/editorial-guidelines.md:
```diff
@@ -148,6 +148,9 @@ These guidelines can be modified through a Pull Request (PR), which the members
 
 * **Display Names:** They SHOULD be used in introductory or explanatory sections where natural language context is more appropriate. In these sections, display names should follow normal text conventions, including spaces between words (e.g., Commitment Discount ID).
 
+* **Normative Keyword Standardization:** To ensure consistency in the specification, normative language MUST use only the BCP-14 keywords MUST, MUST NOT, SHOULD, SHOULD NOT, and MAY. The group agreed the term "RECOMMENDED" MUST NOT be used as a normative keyword. Where legacy text uses RECOMMENDED, it MUST be replaced with SHOULD.
```
```suggestion
* **Normative Keyword Standardization:** To ensure consistency in the specification, normative language MUST use only the BCP-14 keywords MUST, MUST NOT, SHOULD, SHOULD NOT, and MAY. The term "RECOMMENDED" MUST NOT be used as a normative keyword. Where legacy text uses RECOMMENDED, it MUST be replaced with SHOULD.
```

shawnalpay replied:
Committing my own suggestion in order to bring this out of draft.

shawnalpay committed:
Clarify 'RECOMMENDED' usage in requirements

Updated the definition of 'RECOMMENDED' to clarify its future prohibition in normative requirements.

shawnalpay commented on guidelines/normative-requirements-guidelines.md:
```diff
@@ -30,7 +30,7 @@ Grouping and ordering of dataset-level normative requirements ensures clarity, c
   * **MUST NOT** – a prohibition
   * **SHOULD** – recommended but not mandatory
-  * **RECOMMENDED** - recommended but not mandatory (currently used only for presence-related normative requirements, as specified in the [FOCUS Feature Level](#focusfeaturelevel) section)
+  * **RECOMMENDED** - recommended but not mandatory (no longer allowed for use in normative requirements from Dec 2025 onwards; refer to editorial guidelines)
```
```suggestion
  * **RECOMMENDED** - recommended but not mandatory (no longer allowed for use in normative requirements from Dec 2025 onwards; refer to editorial guidelines for more information)
```

shawnalpay commented on guidelines/normative-requirements-guidelines.md:
@ijurica 
There are a couple more spots to touch to scrub the use of the term RECOMMENDED.

ijurica replied:
Addressed and committed the changes directly to the branch

ijurica committed:
FOCUS #1646: normative-requirements-guidelines update (cleanup)

shawnalpay committed:
Update guidelines/editorial-guidelines.md

ijurica commented:
Consider updating the following:

- Specification -> **Introduction** section (/specification/introduction.md file), 
  - row 82 (optional)
  - row 90
- Guidelines -> **Guidelines for writing model rules** section (/guidelines/writing-requirements-model-guidelines.md)
  -  row 37

***Note:*** *I'm not sure if we need to add disclaimers here, similar to the one we added in the Normative Reqs Guidelines.*

---

> ## Introduction
> ...
> ### Typographic Conventions
> 
> The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this specification are to be interpreted as described in [BCP14](https://tools.ietf.org/html/bcp14) [[RFC2119](https://tools.ietf.org/html/rfc2119)][[RFC8174](https://tools.ietf.org/html/rfc8174)] when, and only when, they appear in all capitals, as shown here.
> 
> ### FOCUS Feature Level
> 
> Under each column defined in the FOCUS specification, there exists a 'Feature level' designation that describes the column as 'Mandatory', 'Conditional', or 'Optional'. Feature level is designated based on the following criteria described in the normative requirements in each column definition:
> 
> * If the existence of a column is described with MUST with no conditions of when it applies, then the feature level is designated as 'Mandatory'.
> * If the existence of a column is described as MUST with conditions of when it applies, then the feature level is designated as 'Conditional'.
> * If the existence of a column is described as RECOMMENDED, then the feature level is designated as 'Recommended'.
> * If the existence of a column is described as MAY, then the feature level is designated as 'Optional'.
> ...

---

> ## Guidelines for writing model rules
> ...
> - `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `RECOMMENDED`, `SHOULD`, `MAY`, `OPTIONAL`)
> ...

jpradocueva committed:
Merge branch 'working_draft' into 1646-fr-standardize-normative-language-convert-all-recommended-statements-to-should

jpradocueva committed:
add a note RECOMMENDED is deprecated starting December 2025

jpradocueva commented:
@ijurica, please check if you are okay with the latest changes. The keyword "RECOMMENDED" is deprecated starting December 2025. Also, change from "RECOMMENDED" to "SHOULD" in your latest comment.

shawnalpay commented:
@Matt-Cowsert to check with Sarah (Google) and Vikram (Microsoft) before we proceed with approval.

jpradocueva committed:
remove RECOMMENDED

jpradocueva committed:
Merge branch 'working_draft' into 1646-fr-standardize-normative-language-convert-all-recommended-statements-to-should

Matt-Cowsert commented:
@shawnalpay I have positive confirmation from Google @timwright2000 and Microsoft @vikramdesai01 that we can move forward with the adjustment (over email on 11 and 12 December respectively).

shawnalpay commented on guidelines/writing-requirements-model-guidelines.md:
```diff
@@ -34,7 +34,7 @@ The first stage of conversion of rules from the normative text to model rules is
 - `Condition` - The definition of conditions under which this rule applies
 - `Requirement` - The definition of what is required for model
-- `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `RECOMMENDED`, `SHOULD`, `MAY`, `OPTIONAL`)
+- `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `SHOULD`, `MAY`, `OPTIONAL`)
```
@jpradocueva Are we using `OPTIONAL` anywhere as a normative requirement?  I believe we are not.  Should we therefore strike `OPTIONAL` from the list of allowed values, or do you believe that is outside the scope of this PR?  If the latter, feel free to resolve this suggestion without commitment.
```suggestion
- `Keyword` - The Normative keyword that applies to this rule (Allowed Values: `MUST`, `SHOULD`, `MAY`)
```

shawnalpay replied:
Discussed in Dec 15 Maintainers.  Agreed to strike OPTIONAL.  Committing this suggestion.

shawnalpay committed:
Update guidelines/writing-requirements-model-guidelines.md

jpradocueva commented:
The Member's group approved this PR without objections or abstentions on the December 18 call.

jpradocueva committed:
Merge branch 'working_draft' into 1646-fr-standardize-normative-language-convert-all-recommended-statements-to-should
